### PR TITLE
Checking materialisations before running 'pb run'

### PR DESCRIPTION
**Pull request by cnu1439**
## Description of the change

- [ ] Modified `get_material_names` to return partial pairs so that `pb run` will be skipped for already materialised dates
- [ ] Modified `check_and_generate_more_materials` method to check for existing materialisations for both `auto` and `manual` strategy. Fixing issue related searching newly materialised dates with newly generated/configured feature date 
- [ ] Updated Redshift unit test cases to address new changes.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 1]
```diff
@@ -399,6 +427,11 @@ def get_complete_sequences(sequences: List[Sequence]) -> int:
             prediction_horizon_days,
             inputs,
         )
+
+        # If we want to include partial pairs, we dont need to look for full valid sequences
+        if return_partial_pairs:
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:432]: Should we do an early return even if `len(materials)` is 0 ?

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 2]
```diff
@@ -169,16 +170,38 @@ def _fetch_valid_historic_materials(
...
+        if label_material_name is not None and self.connector.is_valid_table(
+            label_material_name
+        ):
+            found_label_material = True
+
+        # Both not found in the warehouse (no point in including invalid pairs)
+        if not found_feature_material and not found_label_material:
+            return
+
+        # One of them is not found in the warehouse and we don't want to include invalid pairs
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:190]: Can we add some unit tests for this logic

File: tests/unit/RedshiftConnector.py [hunk 3]
```diff
@@ -961,13 +973,19 @@ def test_generate_new_materials_manual_strategy(
...
         result = self.trainer.check_and_generate_more_materials(
             mock_get_material_func,
-            materials=self.materials,
+            materials=self.materials.copy(),
             input_models=self.input_models,
             whtService=self.whtService,
             connector=self.connector,
         )
 
+        print(f"Total call count pb run {mock_rudderpb_run.call_count}")
```
Comment (shekhar-rudder) [inline tests/unit/RedshiftConnector.py:982]: can be removed

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py [hunk 4]
```diff
@@ -407,7 +430,16 @@ def check_and_generate_more_materials(
 
             # Get materials with new feature start date
             # and validate min data requirement again
-            materials = get_material_func(start_date=feature_date)
+
+            # For manual strategy, we need to get the materials for the selected dates separately
+            # because the selected dates may not be in the search window. so creating temporary start and end dates
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/trainers/MLTrainer.py:435]: > so creating temporary start and end dates

Where is this being done ?

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py [hunk 5]
```diff
@@ -393,8 +393,31 @@ def check_and_generate_more_materials(
                     continue
 
             try:
-                for date in [feature_date, label_date]:
-                    whtService.run(feature_package_path, date)
+                # Check wether the materialisation is already present
+                temp_materials = get_material_func(
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/trainers/MLTrainer.py:397]: What does `temp` mean ? Can we rename it to `existing` ?

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py (hunk 1 above)
Comment (cnu1439) [reply src/predictions/rudderstack_predictions/wht/pythonWHT.py:432]: yes, in that case newly generated feature and label dates will get materialized.

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py (hunk 4 above)
Comment (cnu1439) [reply src/predictions/rudderstack_predictions/trainers/MLTrainer.py:435]: This was my previous implementation. realized we can search material registry with generated feature date itself. Will update the comment.

**Review by shekhar-rudder — APPROVED** [review]

**Pull request merged**